src/resources/dac_esp32.cc

```diff
@@ -84,108 +85,103 @@ static dac_channel_t get_dac_channel(int pin) {
  public:
   TAG(DacResource);
-  DacResource(DacResourceGroup* group, dac_channel_t channel);
+  DacResource(ResourceGroup* group, dac_channel_t channel)
+      : Resource(group)
+      , channel_(channel) {}
 
-  virtual ~DacResource() override {
-    unuse_cosine();
-  }
+  virtual ~DacResource() override;
 
   dac_channel_t channel() const { return channel_; }
+  bool uses_cosine() const { return uses_cosine_; }
 
-  esp_err_t use_cosine();
-  esp_err_t unuse_cosine();
+  dac_oneshot_handle_t oneshot_handle() const { return oneshot_handle_; }
+  dac_cosine_handle_t cosine_handle() const { return cosine_handle_; }
+
+  void set_oneshot_handle(dac_oneshot_handle_t handle) {
+    uses_cosine_ = false;
+    oneshot_handle_ = handle;
+  }
+  void set_cosine_handle(dac_cosine_handle_t handle) {
+    uses_cosine_ = true;
+    cosine_handle_ = handle;
+  }
+
+  void release_oneshot();
+  void release_cosine();
 
  private:
   dac_channel_t channel_;
   bool uses_cosine_ = false;
+  // During construction we don't allocate the oneshot/cosine handle. Instead, we
+  // wait for the first output-request to allocate the corresponding handle.
+  union {
+    dac_oneshot_handle_t oneshot_handle_ = null;
+    dac_cosine_handle_t cosine_handle_;
+  };
 };
 
-esp_err_t DacResource::use_cosine() {
-  Locker locker(OS::resource_mutex());
-  esp_err_t err = ESP_OK;
-  if (uses_cosine_) return err;
-  uses_cosine_ = true;
-  cosine_user_count++;
-  if (cosine_user_count == 1) {
-    // First user.
-    err = dac_cw_generator_enable();
+DacResource::~DacResource() {
+  if (uses_cosine_) {
+    release_cosine();
+  } else {
+    release_oneshot();
   }
-  return err;
+  dac_channels_.put(channel_);
 }
 
-esp_err_t DacResource::unuse_cosine() {
-  Locker locker(OS::resource_mutex());
-  esp_err_t err = ESP_OK;
-  if (!uses_cosine_) return err;
-  uses_cosine_ = false;
-  cosine_user_count--;
-  if (cosine_user_count == 0) {
-    // Last user.
-    err = dac_cw_generator_disable();
+void DacResource::release_oneshot() {
+  ASSERT(!uses_cosine());
+
+  auto handle = oneshot_handle_;
+  if (handle != null) {
+    ESP_ERROR_CHECK(dac_oneshot_del_channel(handle));
   }
-  return err;
+  oneshot_handle_ = null;
 }
 
-class DacResourceGroup : public ResourceGroup {
- public:
-  TAG(DacResourceGroup);
-  explicit DacResourceGroup(Process* process)
-      : ResourceGroup(process) {}
-
-  virtual void on_unregister_resource(Resource* resource) override {
-    DacResource* dac_resource = static_cast<DacResource*>(resource);
-
-    dac_output_disable(dac_resource->channel());
+void DacResource::release_cosine() {
+  auto handle = cosine_handle_;
+  if (handle != null) {
+    dac_cosine_stop(handle);
+    ESP_ERROR_CHECK(dac_cosine_del_channel(handle));
   }
- private:
-  static int cosine_user_count_;
-};
-
-DacResource::DacResource(DacResourceGroup* group, dac_channel_t channel)
-    : Resource(group), channel_(channel) {}
-
+  uses_cosine_ = false;
```

Comment:
I'd probably update the `cosine_handle_` before changing `uses_cosine_`, so the invariant is that you only read/write the "correct" union field according to `uses_cosine_`.

Reply:
done.